File: challenge-273/iangoodnight/python/task2.py
"""
## Task 2: B After A

**Submitted by:** [Mohammad Sajid Anwar][0]

You are given a string, `$str`.

Write a script to return `true` if there is at least one `b`, and no `a`
appears after the first `b`.

**Example 1**

```
Input: $str = "aabb"
Output: true
```

**Example 2**

```
Input: $str = "abab"
Output: false
```

**Example 3**

```
Input: $str = "aaa"
Output: false
```

**Example 4**

```
Input: $str = "bbb"
Output: true
```

[0]: https://manwar.org/
"""


def b_after_a(string: str) -> bool:
    """
    Returns `true` if there is at least one `b`, and no `a` appears after the
    first `b`.

    Args:
        string: The input string to check for the presence of `b` and `a`.

    Returns:
        `True` if there is at least one `b`, and no `a` appears after the first
        `b`; otherwise, `False`.
    """
    if "b" not in string:  # fails if there is no `b` in the string
        return False
    b_index = string.index("b")
    # grab a slice of the string after the first `b`
    sliced_after_b = string[b_index + 1:]
    # check if there is no `a` in the slice
    return "a" not in sliced_after_b

File: challenge-273/iangoodnight/typescript/task2.ts
/* ************************************************************************** **
 * ## Task 2: B After A
 *
 * **Submitted by:** [Mohammad Sajid Anwar][2]
 *
 * You are given a string, `$str`.
 *
 * Write a script to return `true` if there is at least one `b`, and no `a`
 * appears after the first `b`.
 *
 * **Example 1**
 *
 * ```
 * Input: $str = "aabb"
 * Output: true
 * ```
 *
 * **Example 2**
 *
 * ```
 * Input: $str = "abab"
 * Output: false
 * ```
 *
 * **Example 3**
 *
 * ```
 * Input: $str = "aaa"
 * Output: false
 * ```
 *
 * **Example 4**
 *
 * ```
 * Input: $str = "bbb"
 * Output: true
 * ```
 *
 * [2]: https://manwar.org/
 * ************************************************************************** */

function bAfterA(str: string): boolean {
  // the regular expression checks for the presence of a `b` followed by any
  // number of characters that are not `a`
  // /         - start of the regular expression
  // ^         - start of the string
  // [^b]*     - any number of characters that are not `b`
  // b         - at least one `b`
  // [^a]*     - any number of characters that are not `a` (even not `b`)
  // $         - end of the string
  // /         - end of the regular expression
  return str.match(/^[^b]*b+[^a]*$/) !== null;
}

export default bAfterA;
